File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;
class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        HashMap<Integer, Integer> oMap = new HashMap<>();
        HashMap<Integer, Integer> yMap = new HashMap<>();
        
        for(int t : topping){
            if(oMap.containsKey(t)){
                oMap.replace(t, oMap.get(t) + 1);
            }else{
                oMap.put(t, 1);
            }
        }
        
        for(int t : topping){
            if(oMap.get(t) == 1){
                oMap.remove(t);
            }else{
                oMap.replace(t, oMap.get(t)-1);
            }
            
            if(yMap.containsKey(t)){
                yMap.replace(t, yMap.get(t) + 1);
            }else{
                yMap.put(t, 1);
            }
            
            if(oMap.size() == yMap.size()){
                answer++;
            }
        }
        
        
        
        return answer;
    }
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    var old = new Map();
    var young = new Map();
    
    for(let i = 0; i < topping.length; i++){
        if(!young.has(topping[i])){
            young.set(topping[i], 0);
        }
        young.set(topping[i], young.get(topping[i]) + 1);
    }
    
    var answer = 0;
    for(let i = 0; i < topping.length; i++){
        if(young.get(topping[i]) === 1){
            young.delete(topping[i]);
        }else{
            young.set(topping[i], young.get(topping[i]) - 1);
        }
        
        if(!old.has(topping[i])){
            old.set(topping[i], 0);
        }
        old.set(topping[i], young.get(topping[i]) + 1);
        
        if(old.size === young.size){
            answer++;
        }
    }
    
    return answer;
}